languages/python/sqlalchemy-oso/sqlalchemy_oso/roles.py

```diff
@@ -1,16 +1,142 @@
+    have a many-to-one (ForeignKey) relationship with this resource model. \
+    A many-to-many relationship to ``user_model`` is added to ``resource_model``; \
+    the relationship is named ``users``.
+
+    :param roles: An order-independent list of the built-in roles for this resource-specific role type.
+    :type roles: List[str]
+
+        .. code-block:: python
+
+            class Team(Base):
+                __tablename__ = "teams"
+
+                id = Column(Integer, primary_key=True)
+                name = Column(String(256))
+
+    :param mutually_exclusive: Boolean flag that sets whether or not users \
+    can have more than one role for a given resource. Defaults to ``True``.
+    :type roles: bool
+
+    :return: the ResourceRole mixin, which must then be mixed into a SQLAlchemy model for the role. E.g.,
+
+        .. code-block:: python
+
+            OrganizationRoleMixin = oso_roles.resource_role_class(
+                Base, User, Organization, ["OWNER", "MEMBER", "BILLING"]
+            )
+
+            class OrganizationRole(Base, OrganizationRoleMixin):
+                pass
+
+
+    """
+    global ROLE_CLASSES
+    ROLE_CLASSES.append(
+        {
+            "user_model": user_model,
+            "resource_model": resource_model,
+        }
+    )
+
+    tablename = f"{resource_model.__name__.lower()}_roles"
+    if mutually_exclusive:
+        unique_constraint = UniqueConstraint(
+            f"{resource_model.__name__.lower()}_id", "user_id"
+        )
+    else:
+        unique_constraint = UniqueConstraint(
+            f"{resource_model.__name__.lower()}_id", "name", "user_id"
+        )
+
+    class ResourceRoleMixin:
+        choices = role_choices
+
+        __tablename__ = tablename
+        id = Column(Integer, primary_key=True)
+        name = Column(String())
+        __table_args__ = (unique_constraint,)
+
+        @validates("name")
+        def validate_name(self, key, name):
+            if name not in self.choices:
+                raise ValueError(
+                    f"{name} Is not a valid choice for {self.__class__.__name__}"
+                )
+            return name
+
+        @declared_attr
+        def user_id(self):
+            type = inspect(user_model).primary_key[0].type
+            name = inspect(user_model).primary_key[0].name
+            table_name = user_model.__tablename__
+            return Column(type, ForeignKey(f"{table_name}.{name}"))
+
+        @declared_attr
+        def user(self):
+            return relationship(user_model.__name__, backref=tablename)
+
+    @declared_attr
+    def resource_id(cls):
+        type = inspect(resource_model).primary_key[0].type
+        name = inspect(resource_model).primary_key[0].name
+        table_name = resource_model.__tablename__
+        return Column(type, ForeignKey(f"{table_name}.{name}"))
+
+    @declared_attr
+    def resource(cls):
+        return relationship(resource_model.__name__, backref="roles")
```

Comment:
What about if there are multiple role models for the resource? Then the backref would be ambiguous

Reply:
Realized this as well. I'm trying to think about whether it makes sense to support multiple roles types for the same resource. If not, we might want to just throw an error if someone tries to do that.

Reply:
I updated `resource_role_class()` to throw an error for now. In the future we can think about if we need to support this.

Reply:
The use case might be having roles for users and groups, so treating a group as an actor type.